### ISSUE TITLE
Texture from IMG element does not use actual dimensions of image

### ISSUE THREAD

**Issue by mlantin** [processing/p5.js#5668]
### Most appropriate sub-area of p5.js?

- [ ] Accessibility (Web Accessibility)
- [ ] Build tools and processes
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [X] DOM
- [ ] Events
- [ ] Friendly error system
- [X] Image
- [ ] IO (Input/Output)
- [ ] Localization
- [ ] Math
- [ ] Unit Testing
- [ ] Typography
- [ ] Utilities
- [X] WebGL
- [ ] Other (specify if possible)

### p5.js version

1.4.1

### Web browser and version

Version 1.36.112 Chromium: 99.0.4844.51 (Official Build) (x86_64)

### Operating System

MacOS 12.3.1

### Steps:
1. Use a file input to upload an image of power-of-two size (like 1024x1024) into an existing \<img\> element that has a non-power-of-two size. Like 40x40.
2. Use that \<img\> element as a texture with a texture wrap mode of REPEAT.
3. p5js warns that the image is not a power of two and that wrap mode will be changed to CLAMP.

Problem seems to be that it's using the width and height attributes of the \<img\> element and should be using the naturalWidth and naturalHeight attributes.

**Comment by flargus** [comment]
Hey, I'm having a similar issue with textureWrap()
Did you find a workaround to this problem?

**Comment by mlantin** [comment]
No I didn't find an elegant workaround. I ended up loading the image twice (once through the img element and once with loadimage). I did look through the p5.js code to see if it would be a simple fix. It's not. The width/height are used quite generically so it would be a bigger fix to suss out when to use naturalWidth/Height instead. I'm neck deep in a project right now so I can't volunteer to fix it (yet).

**Comment by davepagurek** [comment]
I just double checked and confirmed that if we override the existing `p5.Texture` functions and make them use `naturalWidth` and `naturalHeight`, then [textures successfully wrap.](https://editor.p5js.org/davepagurek/sketches/IXgpMfbO-) So I agree that the fix for this would be to update [the constructor](https://github.com/processing/p5.js/blob/ea0fa1d270003437c2750da0438d48e93954ea02/src/webgl/p5.Texture.js#L92) and [the `update` function](https://github.com/processing/p5.js/blob/ea0fa1d270003437c2750da0438d48e93954ea02/src/webgl/p5.Texture.js#L188) to check `naturalWidth` and `naturalHeight` first if it's an image element, and fall back to `width`/`height` otherwise.

**Comment by leslieyip02** [comment]
Hi, I took a look at the demo and it seems that `this.width` and `this.height` are being set to `undefined`. It seems that setting the width and height to any power of 2 will cause the texture to wrap properly.

> I just double checked and confirmed that if we override the existing `p5.Texture` functions and make them use `naturalWidth` and `naturalHeight`, then [textures successfully wrap.](https://editor.p5js.org/davepagurek/sketches/IXgpMfbO-) So I agree that the fix for this would be to update [the constructor](https://github.com/processing/p5.js/blob/ea0fa1d270003437c2750da0438d48e93954ea02/src/webgl/p5.Texture.js#L92) and [the `update` function](https://github.com/processing/p5.js/blob/ea0fa1d270003437c2750da0438d48e93954ea02/src/webgl/p5.Texture.js#L188) to check `naturalWidth` and `naturalHeight` first if it's an image element, and fall back to `width`/`height` otherwise.

**Comment by davepagurek** [comment]
@leslieyip02 thanks for catching that! My mistake, the inside of the branch was missing the `elt.` prefix. I've updated the sketch to now have this:

```js
if (this.src.elt && this.src.elt.naturalWidth && this.src.elt.naturalHeight) {
  this.width = this.src.elt.naturalWidth
  this.height = this.src.elt.naturalHeight
}
```

> It seems that setting the width and height to any power of 2 will cause the texture to wrap properly.

Setting the width/height to any power of 2 will just bypass p5's check before asking WebGL to repeat the texture. It seems that browsers can still support REPEAT for all sizes sometimes. [MDN says this though, about the spec in general:](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Tutorial/Using_textures_in_WebGL)

> if the texture is a power of 2 in both dimensions then WebGL can do higher quality filtering, it can use mipmap, and it can set the wrapping mode to REPEAT or MIRRORED_REPEAT.

But based on that description, it sounds like the spec only *requires* this to work for power of 2 sizes, so it's probably best to just keep p5's check in place anyway, just to be safe, but basing it on `this.src.elt.naturalWidth`.

**Comment by kate-grant** [comment]
I saw the Good First Issue label and tried implementing the change suggested to the power of 2 check but found that this.src.elt.naturalWidth didn't seem to exist--I would love to work on a PR if possible.

**Comment by kate-grant** [comment]
@davepagurek I had to solve this in a different way than the original recommended solution. The original solution made changes to the constructor, changing `this.width` and `this.height`, but this interfered with the update function in some cases [`data.setModified(true);`](https://github.com/processing/p5.js/blob/main/src/webgl/p5.Texture.js#L207). Instead I utilized this logic in `setWrapMode` with new variables `wrapWidth` and `wrapHeight` for use in the `isPowerOfTwo` check. Let me know if you think this is ok.

I added two unit tests for textureWrap REPEAT on textures from image elements for power-of-two and non-power-of-two (the power-of-two case should allow REPEAT and the non-power-of-two case should set to CLAMP). I had to add two things to the Texture unit test setup fn--let me know if this is ok.
